### PR TITLE
Fix broken ComparisonTable columns on screen readers

### PR DESCRIPTION
**Pull request by joshfarrant**
## Summary

Removed the `aria-hidden` attribute from empty `<th>` cells which was causing table structure to be misrepresented by screen readers

## List of notable changes:

- Removed `aria-hidden` from `ComparisonTable` column headers

## What should reviewers focus on?

- Ensure the table structure is accurately represented when using a screen reader

## Steps to test:

1. Open [a ComparisonTable example on Storybook](https://primer-52bb15c8a0-26139705.drafts.github.io/brand/storybook/?path=/story/components-comparisontable--minimal).
1. Enable a screen reader (eg VoiceOver).
1. Navigate through the table and ensure that column numbers and titles are accurately communicated.

## Supporting resources (related issues, external links, etc):

- https://github.com/github/primer/issues/3753

## Contributor checklist:

- [x] All new and existing CI checks pass
- [x] Tests prove that the feature works and covers both happy and unhappy paths
- [x] Any drop in coverage, breaking changes or regressions have been documented above
- [x] New visual snapshots have been generated / updated for any UI changes
- [x] All developer debugging and non-functional logging has been removed
- [x] Related issues have been referenced in the PR description

## Reviewer checklist:

- [ ] Check that pull request and proposed changes adhere to our [contribution guidelines](../../CONTRIBUTING.md) and [code of conduct](../../CODE_OF_CONDUCT.md)
- [ ] Check that tests prove the feature works and covers both happy and unhappy paths
- [ ] Check that there aren't other open Pull Requests for the same update/change

## Videos

Note the mis-reported column numbers in the first video.

<table>
<tr>
<th> Before</th> <th> After </th>
</tr>
<tr>
<td valign="top">

https://github.com/user-attachments/assets/90476b0d-0aa9-403c-bb84-1de472be2084

 </td>
<td valign="top">

https://github.com/user-attachments/assets/8e4d783c-4b10-45eb-a1b5-49eca2f2c575

</td>
</tr>
</table>

### CODE REVIEW

File: packages/react/src/ComparisonTable/ComparisonTable.tsx [hunk 1]
```diff
@@ -55,8 +55,7 @@ export const _ComparisonTable = forwardRef(
             children: React.Children.map(child.props.children, (rowChild, rowChildIndex) => {
               if (rowChild.type === Cell) {
                 return React.cloneElement(rowChild, {
-                  as: 'th',
-                  'aria-hidden': !rowChild.props.children ? 'true' : undefined,
+                  as: rowChild.props.children ? 'th' : 'td',
```
Comment (rezrah) [inline packages/react/src/ComparisonTable/ComparisonTable.tsx:58]: Is the conditional behaviour needed here? Wouldn't it always be a `th` for the first row?

Comment (rezrah) [reply packages/react/src/ComparisonTable/ComparisonTable.tsx:58]: Column alignment is also off for the first row... not sure if related to this or the removal of aria-hidden

[before](https://stunning-chainsaw-j82glqz.pages.github.io/?path=/story/components-comparisontable--minimal)
Screenshot 2024-08-15 at 17 35 45


[after](https://primer-52bb15c8a0-26139705.drafts.github.io/brand/storybook/?path=/story/components-comparisontable--minimal)
Screenshot 2024-08-15 at 17 35 52

File: packages/react/src/ComparisonTable/ComparisonTable.visual.spec.ts-snapshots/Visual-Comparison-ComparisonTable-ComparisonTable-Minimal-1-linux.png
Comment (rezrah) [inline packages/react/src/ComparisonTable/ComparisonTable.visual.spec.ts-snapshots/Visual-Comparison-ComparisonTable-ComparisonTable-Minimal-1-linux.png:1]: Related to https://github.com/primer/brand/pull/694/files#r1718673566. You can see the column misalignment in this diff 👇

File: .changeset/great-houses-argue.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@primer/react-brand': patch
+---
+
+Fixed issue where `ComparisonTable` column structure was broken on screen readers.
```
Comment (rezrah) [inline .changeset/great-houses-argue.md:5]: To be a little more specific on what is broken...

```suggestion
Improved screen reader accuracy of column lengths in `ComparisonTable`. 
```

File: packages/react/src/ComparisonTable/ComparisonTable.tsx (hunk 1 above)
Comment (joshfarrant) [reply packages/react/src/ComparisonTable/ComparisonTable.tsx:58]: Sorry, I should have added context on the `th`/`td` behaviour here (and checked the snapshots after making the change!).

I added this as axe was complaining about empty `th` cells. You can see the rule here

https://dequeuniversity.com/rules/axe/4.7/empty-table-header?application=axeAPI

I couldn't find a recommended approach for this kind of 2-way table where the top-left cell doesn't have any meaning, hence why I landed on the `td` approach.

Thanks for spotting the CSS issues. Fixing the CSS might take a bit of extra work as there are a lot of selectors which expect that first cell to be a `th`.

Let me dig into this a bit more and find an approach that works.

Comment (joshfarrant) [reply packages/react/src/ComparisonTable/ComparisonTable.tsx:58]: Sorry, I should have added context on the `th`/`td` behaviour here (and checked the snapshots after making the change!).

I added this as axe was complaining about empty `th` cells. You can see the rule here

https://dequeuniversity.com/rules/axe/4.7/empty-table-header?application=axeAPI

I couldn't find a recommended approach for this kind of 2-way table where the top-left cell doesn't have any meaning, hence why I landed on the `td` approach.

I've reached out to Tetralogical to see if they have a recommended approach in this scenario

File: packages/react/src/recipes/SolutionTemplates/SolutionPage/UseCase/UseCase.visual.spec.ts-snapshots/Visual-Comparison-UseCase-UseCase-Minimum-Light-1-linux.png
Comment (rezrah) [inline packages/react/src/recipes/SolutionTemplates/SolutionPage/UseCase/UseCase.visual.spec.ts-snapshots/Visual-Comparison-UseCase-UseCase-Minimum-Light-1-linux.png:1]: Can't for life of me figure out what's changed here 😄 ... any ideas?

Comment (joshfarrant) [reply packages/react/src/recipes/SolutionTemplates/SolutionPage/UseCase/UseCase.visual.spec.ts-snapshots/Visual-Comparison-UseCase-UseCase-Minimum-Light-1-linux.png:1]: No me neither — I've had this a few times recently, I don't know if it's related to me running the snapshots locally vs using the label

File: packages/react/src/ComparisonTable/ComparisonTable.tsx (hunk 1 above)
Comment (joshfarrant) [reply packages/react/src/ComparisonTable/ComparisonTable.tsx:58]: Based on [the suggestion from Tetralogical](https://github.com/github/primer/issues/3753#issuecomment-2293478584) I've opted to keep this as a `<td>` and have fixed the style issues

File: packages/react/src/recipes/SolutionTemplates/SolutionPage/UseCase/UseCase.visual.spec.ts-snapshots/Visual-Comparison-UseCase-UseCase-Minimum-Light-1-linux.png
Comment (rezrah) [reply packages/react/src/recipes/SolutionTemplates/SolutionPage/UseCase/UseCase.visual.spec.ts-snapshots/Visual-Comparison-UseCase-UseCase-Minimum-Light-1-linux.png:1]: Worth removing these changes, or fixing it in a follow. up. FWIW I usually trust my local machine to regen vs Actions...

**Review by rezrah — APPROVED** [review]
Looks great, thanks for fixing this @joshfarrant

**Pull request merged**